programs/bpf_loader/src/serialization.rs

```diff
@@ -2,17 +2,78 @@ use {
     byteorder::{ByteOrder, LittleEndian, WriteBytesExt},
     solana_rbpf::{aligned_memory::AlignedMemory, ebpf::HOST_ALIGN},
     solana_sdk::{
-        bpf_loader_deprecated,
+        account::{Account, AccountSharedData},
+        bpf_loader, bpf_loader_deprecated,
         entrypoint::{BPF_ALIGN_OF_U128, MAX_PERMITTED_DATA_INCREASE},
         instruction::InstructionError,
         pubkey::Pubkey,
         system_instruction::MAX_PERMITTED_DATA_LENGTH,
-        transaction_context::{InstructionContext, TransactionContext},
+        transaction_context::{InstructionAccount, InstructionContext, TransactionContext},
     },
-    std::{io::prelude::*, mem::size_of},
+    std::{collections::hash_map::Entry, collections::HashMap, io::prelude::*, mem::size_of},
 };
 
-/// Look for a duplicate account and return its position if found
+/// Holds information about duplicate instruction accounts. It can be used as a
+/// cache to avoid calculating duplicates repeatedly.
+struct DuplicateInstructionAccounts<'a> {
+    instruction_context: &'a InstructionContext,
+    /// Used to translate between position of duplicate and index in instruction.
+    index_offset: usize,
+    /// Maps an instruction account's index_in_transaction to its position.
+    map: HashMap<usize, usize>,
+}
+
+impl<'a> DuplicateInstructionAccounts<'a> {
+    fn new(instruction_context: &'a InstructionContext) -> Result<Self, InstructionError> {
+        // We create the HashMap with capacity to avoid reallocations. If there
+        // are duplicate instruction accounts, the map's capacity will not be
+        // used up. Neglect this as elements are small and the number of
+        // accounts is bounded.
+        let mut map: HashMap<usize, usize> =
+            HashMap::with_capacity(instruction_context.get_number_of_instruction_accounts());
```

Comment:
Could you try using a vec instead to avoid the hashing that comes with using a hashmap? It should be more memory efficient as well.